crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -57,279 +101,522 @@ impl PackageWatcher {
         })
     }
 
-    pub async fn get_package_data(&self) -> Vec<WorkspaceData> {
-        self.package_data
-            .lock()
-            .expect("not poisoned")
-            .clone()
-            .into_values()
-            .collect()
+    /// Get the package data. If the package data is not available, this will
+    /// block until it is.
+    pub async fn wait_for_package_data(
+        &self,
+    ) -> Result<Vec<WorkspaceData>, watch::error::RecvError> {
+        let mut recv = self.package_data.clone();
+        recv.get().await.map(|v| v.values().cloned().collect())
     }
 
-    pub async fn get_package_manager(&self) -> PackageManager {
-        *self.manager_rx.borrow()
+    /// A convenience wrapper around `FutureExt::now_or_never` to let you get
+    /// the package data if it is immediately available.
+    pub fn get_package_data(&self) -> Option<Result<Vec<WorkspaceData>, watch::error::RecvError>> {
+        let mut recv = self.package_data.clone();
+        let data = if let Some(Ok(inner)) = recv.get().now_or_never() {
+            Some(Ok(inner.values().cloned().collect()))
+        } else {
+            None
+        };
+        data
+    }
+
+    /// Get the package manager. If the package manager is not available, this
+    /// will block until it is.
+    pub async fn wait_for_package_manager(
+        &self,
+    ) -> Result<PackageManager, watch::error::RecvError> {
+        let mut recv = self.manager_rx.clone();
+        recv.get().await.map(|s| s.manager)
+    }
+
+    /// A convenience wrapper around `FutureExt::now_or_never` to let you get
+    /// the package manager if it is immediately available.
+    pub fn get_package_manager(&self) -> Option<Result<PackageManager, watch::error::RecvError>> {
+        let mut recv = self.manager_rx.clone();
+        // the borrow checker doesn't like returning immediately here so assign to a var
+        let data = if let Some(Ok(inner)) = recv.get().now_or_never() {
+            Some(Ok(inner.manager))
+        } else {
+            None
+        };
+        data
+    }
+
+    pub fn watch(&self) -> OptionalWatch<HashMap<AbsoluteSystemPathBuf, WorkspaceData>> {
+        self.package_data.clone()
     }
 }
 
 /// The underlying task that listens to file system events and updates the
 /// internal package state.
 struct Subscriber<T: PackageDiscovery> {
-    exit_rx: oneshot::Receiver<()>,
-    filter: WorkspaceGlobs,
-    recv: broadcast::Receiver<Result<Event, NotifyError>>,
+    /// we need to hold on to this. dropping it will close the downstream
+    /// data dependencies
+    _recv_tx: Arc<watch::Sender<Option<broadcast::Receiver<Result<Event, NotifyError>>>>>,
+
+    recv: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+    backup_discovery: Arc<AsyncMutex<T>>,
```

Comment:
Is this mutable? Why does it need a lock?

Reply:
fixed